src/three-components/PMREMGenerator.ts

```diff
@@ -0,0 +1,310 @@
+
+const $setup = Symbol('setup');
+const $generateMipmaps = Symbol('generateMipmaps');
+const $packMipmaps = Symbol('packMipmaps');
+const $addLodMeshes = Symbol('addLodMeshes');
+const $updateLodMeshes = Symbol('updateLodMeshes');
+
+/**
+ * This class generates a Prefiltered, Mipmapped Radiance Environment Map
+ * (PMREM) from a cubeMap environment texture. This allows different levels of
+ * blur to be quickly accessed based on material roughness. It is packed into a
+ * special CubeUV format that allows us to perform custom interpolation so that
+ * we can support nonlinear formats such as RGBE.
+ */
+export default class PMREMGenerator {
+  private mipmapShader = new MipmapShader();
+  private packingShader = new PackingShader();
+  private cubeCamera = new CubeCamera(0.1, 100, 1);
+  private flatCamera = new OrthographicCamera(-1, 1, 1, -1);
+  private mipmapScene = new Scene();
+  private packingScene = new Scene();
+  private boxMesh = new Mesh(new BoxBufferGeometry(), this.mipmapShader);
+  private plane = new PlaneBufferGeometry(1, 1);
+  // Hard-coded to max faceSize = 256 until we can add a uniform.
+  private maxLods = 8;
+  private cubeLods: Array<WebGLRenderTargetCube>;
+  private cubeUVRenderTarget: WebGLRenderTarget|null = null;
+  private meshes: Array<Mesh>;
+
+  constructor(private renderer: WebGLRenderer) {
+    (this.boxMesh.material as Material).side = DoubleSide;
+    this.mipmapScene.add(this.boxMesh);
+    this.cubeLods = [];
+    this.meshes = [];
+  }
+
+  update(cubeTarget: WebGLRenderTargetCube): WebGLRenderTarget {
+    this[$setup](cubeTarget);
+
+    const {renderer} = this;
+    const {gammaInput, gammaOutput, toneMapping, toneMappingExposure} =
+        renderer;
+    const currentRenderTarget = renderer.getRenderTarget();
+
+    renderer.toneMapping = LinearToneMapping;
+    renderer.toneMappingExposure = 1.0;
+    renderer.gammaInput = false;
+    renderer.gammaOutput = false;
+
+    this[$generateMipmaps](cubeTarget);
+    this[$packMipmaps]();
+
+    renderer.setRenderTarget(currentRenderTarget);
+    renderer.toneMapping = toneMapping;
+    renderer.toneMappingExposure = toneMappingExposure;
+    renderer.gammaInput = gammaInput;
+    renderer.gammaOutput = gammaOutput;
+
+    return this.cubeUVRenderTarget!;
+  }
+
+  private[$setup](cubeTarget: WebGLRenderTargetCube) {
+    const params = {
+      format: cubeTarget.texture.format,
+      magFilter: NearestFilter,
+      minFilter: NearestFilter,
+      type: cubeTarget.texture.type,
+      generateMipmaps: false,
+      anisotropy: cubeTarget.texture.anisotropy,
+      encoding: cubeTarget.texture.encoding
+    };
+
+    const {maxLods, flatCamera} = this;
+
+    // Math.log(cubeTarget.width) / Math.log(2) - 2;  // IE11 doesn't support
+    // Math.log2
+
+    let offset = 0;
+    for (let i = 0; i <= maxLods; i++) {
+      const sizeLod = Math.pow(2, i);
+      const renderTarget = new WebGLRenderTargetCube(sizeLod, sizeLod, params);
```

Comment:
Do we create an extra `WebGLRenderTargetCube` here and throw it away in the case that `i === maxLods`?

Reply:
indeed, fixed.